core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,492 @@
+        if (currentState != TRANSIENT_FAILURE) {
+          picker = new Picker(PickResult.withNoResult());
+          updateBalancingState(CONNECTING, picker);
+        }
+        break;
+      case READY:
+        picker = new Picker(PickResult.withSubchannel(subchannel));
+        updateBalancingState(READY, picker);
+        shutdownRemaining(subchannel.getAddresses().getAddresses().get(0), subchannel);
+        break;
+      case TRANSIENT_FAILURE:
+        // If we are looking at current channel ...
+        if (subchannels.get(addressIndex.getCurrentAddress()).equals(subchannel)) {
+          // ... and it is the last address, report TRANSIENT_FAILURE.
+          // Only happens once per address update.
+          addressIndex.increment();
+          if (enterTransientFailure()) {
+            iterate = false;
+            addressIndex.reset();
+            helper.refreshNameResolution();
+            picker = new Picker(PickResult.withError(stateInfo.getStatus()));
+            updateBalancingState(TRANSIENT_FAILURE, picker);
+            // If we are not at the last address, try a connection attempt to the next address.
+          } else {
+            requestConnection();
+          }
+          // If we are looking at another address that retried after backoff
+          // and entered TRANSIENT_FAILURE, we do not do any incrementing logic,
+          // as individual subchannels will perform their own backoffs.
+        }
+        break;
+      default:
+        throw new IllegalArgumentException("Unsupported state:" + newState);
+    }
+  }
+
+  private void updateBalancingState(ConnectivityState state, SubchannelPicker picker) {
+    if (state != currentState || currentState != CONNECTING) {
+      currentState = state;
+      helper.updateBalancingState(state, picker);
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    for (Subchannel subchannel : subchannels.values()) {
+      subchannel.shutdown();
+      states.put(subchannel, null);
+    }
+    subchannels.clear();
+  }
+
+  /**
+  * Shuts down remaining subchannels. Called when a subchannel becomes ready, which means
+  * that all other subchannels must be shutdown.
+  */
+  private void shutdownRemaining(SocketAddress activeAddr, Subchannel activeSubchannel) {
+    for (Subchannel subchannel : subchannels.values()) {
+      if (!subchannel.equals(activeSubchannel)) {
+        subchannel.shutdown();
+      }
+    }
+    iterate = true;
+    subchannels.clear();
+    states.clear();
+    states.put(activeSubchannel, READY);
+    subchannels.put(activeAddr, activeSubchannel);
+  }
+
+  /**
+  * Requests a connection to the next applicable address/subchannel.
+  * If the current channel has already attempted a connection, we attempt a connection
+  * to the next address/subchannel in our list.
+  */
+  @Override
+  public void requestConnection() {
+    if (subchannels.size() == 0) {
+      return;
+    }
+
```

Comment:
A refactor idea to make the code easier to read is:
```
if addressIndex is valid then:
  subchannel = map.getOrDefault(address, new Subchannel())
  if subchannel.state is in IDLE: subchannel.requestConnection()
  if subchannel.state is in TF: increment and recursively requestConnection()
  // worth discussion whether we need to recursively request connection on already CONNECTING subchannel, in 
  // happy eyeball maybe but in pickfirst maybe not.
```

And actually a better solution here is iteration rather than recursion, because its easier to understand.

Reply:
Resolved in meeting: Use `computeIfAbsent()` to simplify logic.